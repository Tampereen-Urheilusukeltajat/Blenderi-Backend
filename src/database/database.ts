import config from '../../knexfile';
import knex from 'knex';

export enum Env {
  development = 'development',
  production = 'production',
}

const ENV = (process.env.ENVIRONMENT as Env) ?? Env.development;

export const knexController = knex(config[ENV]);
