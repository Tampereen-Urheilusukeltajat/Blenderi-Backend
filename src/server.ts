import fastify, { FastifyInstance } from 'fastify';
import { fastifySwagger } from '@fastify/swagger';
import { fastifyHelmet } from '@fastify/helmet';
import fastifyCors from '@fastify/cors';
import { fastifyAutoload } from '@fastify/autoload';
import { TypeBoxTypeProvider } from '@fastify/type-provider-typebox'

import { log } from './lib/log';
import path from 'path';

export const buildServer = async (opts: {
  routePrefix: string
}): Promise<FastifyInstance> => {
  const server = fastify({
    logger: false,
    ignoreTrailingSlash: true,
    ajv: {
      customOptions: {
        removeAdditional: 'all' // Remove additional params from the body etc
      }
    }
  });

  // Register plugins and routes
  server
    .addSchema({
      $id: 'error',
      type: 'object',
      properties: {
        statusCode: { type: 'number' },
        error: { type: 'string' },
        message: { type: 'string' }
      }
    })
    .register(fastifySwagger, {
      routePrefix: `${opts.routePrefix}/documentation`,
      swagger: {
        info: {
          title: 'Blenderi REST API',
          description: 'Documentation for Blenderi REST API',
          version: '0.0.1'
        },
        securityDefinitions: {
          bearerAuth: {
            type: 'apiKey',
            name: 'Authorization',
            in: 'header'
          }
        },
        security: [{
          bearerAuth: []
        }],
        consumes: ['serverlication/json'],
        produces: ['serverlication/json'],
        tags: [{
          name: 'Utility',
          description: 'Utility endpoints'
        }
        ]
      },
      exposeRoute: true
    })
    .register(fastifyHelmet, {
      contentSecurityPolicy: {
        directives: {
          defaultSrc: ['\'self\''],
          styleSrc: ['\'self\'', '\'unsafe-inline\''],
          imgSrc: ['\'self\'', 'data:', 'validator.swagger.io'],
          scriptSrc: ['\'self\'', 'https: \'unsafe-inline\'']
        }
      }
    })
    .register(fastifyCors, {
      origin: (origin, cb) => {
        // TODO: Configure CORS
        cb(null, true);
      }
    })
    .register(fastifyAutoload, {
      dir: path.join(__dirname, 'routes'),
      dirNameRoutePrefix: (_folderParent, folderName) => `${opts.routePrefix}/${folderName}`
    })
    .setErrorHandler(async (error, request, reply) => {
      log.error({
        error: error.name,
        message: error.message,
        url: request.url,
        method: request.method,
        body: request.body,
        stack: error.stack
      });

      await reply.status(500).send({
        statusCode: 500,
        error: 'Internal Server Error',
        message: 'Internal Server Error'
      });
    })
    .withTypeProvider<TypeBoxTypeProvider>();

  return server;
};
