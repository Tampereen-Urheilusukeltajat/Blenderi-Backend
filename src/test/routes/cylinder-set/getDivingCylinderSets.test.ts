import { describe, test, expect, beforeAll, afterAll } from '@jest/globals';
import { FastifyInstance } from 'fastify';
import {
  createTestDatabase,
  dropTestDatabase,
  startRedisConnection,
  stopRedisConnection,
} from '../../../lib/utils/testUtils';
import { knexController } from '../../../database/database';
import { buildServer } from '../../../server';

describe('get cylinder sets', () => {
  const getTestInstance = async (): Promise<FastifyInstance> =>
    buildServer({
      routePrefix: 'api',
    });

  beforeAll(async () => {
    await createTestDatabase('get_diving_cylinder_sets');
    await startRedisConnection();
  });

  afterAll(async () => {
    await dropTestDatabase();
    await knexController.destroy();
    await stopRedisConnection();
  });

  let server;
  let headers: object;
  test('returns users diving cylinder sets', async () => {
    server = await getTestInstance();
    const authRes = await server.inject({
      url: '/api/login',
      method: 'POST',
      payload: {
        email: 'user@taursu.fi',
        password: 'salasana',
      },
    });
    const tokens = JSON.parse(authRes.body);
    headers = { Authorization: 'Bearer ' + String(tokens.accessToken) };

    const res = await server.inject({
      url: `api/cylinder-set/?userId=a59faf66-4f75-11ed-98ae-77941df77788`,
      method: 'GET',
      headers,
    });

    expect(res.statusCode).toEqual(200);

    const body = JSON.parse(res.body);
    expect(body.length).toEqual(3);
    expect(body).toMatchSnapshot();
  });

  test('does not allow reading other users diving cylinders', async () => {
    server = await getTestInstance();
    const authRes = await server.inject({
      url: '/api/login',
      method: 'POST',
      payload: {
        email: 'user@taursu.fi',
        password: 'salasana',
      },
    });
    const tokens = JSON.parse(authRes.body);
    headers = { Authorization: 'Bearer ' + String(tokens.accessToken) };

    const res = await server.inject({
      url: `api/cylinder-set/?userId=a59faf66-4f75-11ed-98ae-77941df77789`,
      method: 'GET',
      headers,
    });

    expect(res.statusCode).toEqual(403);
  });

  test('admin is able to read other users diving cylinders', async () => {
    server = await getTestInstance();
    const authRes = await server.inject({
      url: '/api/login',
      method: 'POST',
      payload: {
        email: 'admin@taursu.fi',
        password: 'salasana',
      },
    });
    const tokens = JSON.parse(authRes.body);
    headers = { Authorization: 'Bearer ' + String(tokens.accessToken) };

    const dcsResponse = await server.inject({
      url: `api/cylinder-set/?userId=a59faf66-4f75-11ed-98ae-77941df77788`,
      method: 'GET',
      headers,
    });

    expect(dcsResponse.statusCode).toEqual(200);

    const body = JSON.parse(dcsResponse.body);
    expect(body.length).toEqual(3);
  });

  test('returns empty array if user does not have cylinder sets', async () => {
    server = await getTestInstance();
    const authRes = await server.inject({
      url: '/api/login',
      method: 'POST',
      payload: {
        email: 'admin@taursu.fi',
        password: 'salasana',
      },
    });
    const tokens = JSON.parse(authRes.body);
    headers = { Authorization: 'Bearer ' + String(tokens.accessToken) };

    const dcsResponse = await server.inject({
      url: `api/cylinder-set/?userId=a59faf66-4f75-11ed-98ae-77941df77790`,
      method: 'GET',
      headers,
    });

    expect(dcsResponse.statusCode).toEqual(200);

    const body = JSON.parse(dcsResponse.body);
    expect(body).toEqual([]);
  });

  test('returns empty array if user is not found', async () => {
    server = await getTestInstance();
    const authRes = await server.inject({
      url: '/api/login',
      method: 'POST',
      payload: {
        email: 'admin@taursu.fi',
        password: 'salasana',
      },
    });
    const tokens = JSON.parse(authRes.body);
    headers = { Authorization: 'Bearer ' + String(tokens.accessToken) };

    const dcsResponse = await server.inject({
      url: `api/cylinder-set/?userId=a59faf66-4f75-11ed-98ae-77941df77799`,
      method: 'GET',
      headers,
    });

    expect(dcsResponse.statusCode).toEqual(200);

    const body = JSON.parse(dcsResponse.body);
    expect(body).toEqual([]);
  });
});
