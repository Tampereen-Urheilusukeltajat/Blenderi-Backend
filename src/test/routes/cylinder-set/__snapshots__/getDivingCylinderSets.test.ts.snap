// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get cylinder sets returns users diving cylinder sets 1`] = `
[
  {
    "cylinders": [
      {
        "id": "a62ab1b8-4f88-11ed-aa8f-a7bc5ca309a3",
        "inspection": "2010-11-11T00:00:00.000Z",
        "material": "carbon fiber",
        "pressure": 200,
        "serialNumber": "aa",
        "volume": 10,
      },
    ],
    "id": "a4e1035e-f36e-4056-9a1b-5925a3c5793e",
    "name": "aa",
    "owner": "a59faf66-4f75-11ed-98ae-77941df77788",
  },
  {
    "cylinders": [
      {
        "id": "b62ab1b8-4f88-11ed-aa8f-a7bc5ca309a3",
        "inspection": "2010-11-11T00:00:00.000Z",
        "material": "carbon fiber",
        "pressure": 200,
        "serialNumber": "bb",
        "volume": 10,
      },
    ],
    "id": "b4e1035e-f36e-4056-9a1b-5925a3c5793e",
    "name": "bb",
    "owner": "a59faf66-4f75-11ed-98ae-77941df77788",
  },
  {
    "cylinders": [
      {
        "id": "1e54c95c-c2fe-4d86-9406-c88f45c0bde9",
        "inspection": "2010-11-11T00:00:00.000Z",
        "material": "steel",
        "pressure": 200,
        "serialNumber": "koodi",
        "volume": 10,
      },
      {
        "id": "362ab1b8-4f88-11ed-aa8f-a7bc5ca309a3",
        "inspection": "2010-11-11T00:00:00.000Z",
        "material": "carbon fiber",
        "pressure": 200,
        "serialNumber": "toinenkoodi",
        "volume": 10,
      },
    ],
    "id": "f4e1035e-f36e-4056-9a1b-5925a3c5793e",
    "name": "pullonnimi",
    "owner": "a59faf66-4f75-11ed-98ae-77941df77788",
  },
]
`;
